text/router-enhancements.md

```diff
@@ -0,0 +1,211 @@
+piece of code that the dynamic segment has to validate in order of that route to match.
+
+In the previous example the constraint would be that the scale is one of our white-listed values, and
+any other would cause the route to miss.
+
+
+# Detailed design
+
+
+### Optional segments.
+
+Mimicking rails micro syntax, I propose optional segments to be enclosed between parenthesis. The
+example of the Apple store would be encoded
+
+```js
+Router.map(function() {
+  this.route('product', { path: '(/:region)/:productName' });
+});
+
+
+router.serialize('/mac'); // => routeName='products', params: { productName: 'mac' }
+router.serialize('/uk/mac'); // => routeName='products', params: { productName: 'mac', region: 'uk' }
+```
+
+### Dynamic segment's constraints
+
+Also stealing Ruby on Rails' syntax, the user can provide a constraint using the
+config object passed as second argument.
+
+
+```js
+const allowedScales = ['celsius', 'fahrenheit', 'kelvin'];
+
+function isAllowedScale(scaleName) {
+  return allowedScales.includes(scaleName);
+}
+
+Router.map(function() {
+  this.route('forecast', { path: 'forecast/:scale', constraints: { scale: isAllowedScale } });
+});
+
+
+router.serialize('/forecast/kelvin'); // => routeName='forecast', params: { scale: 'kelvin' }
+router.serialize('/forecast/newton'); // => not recognized route.
+```
+
+Although the generic case of receiving a function that returns a boolean is the most generic one,
+it might worth to add a RegExp support as a convenience.
+
+```js
+Router.map(function() {
+  this.route('forecast', { path: 'forecast/:scale', constraints: { scale: /(celsius|fahrenheit|kelvin)/ } });
+
+  // Equivalent to:
+  // this.route('forecast', {
+  //   path: 'forecast/:scale',
+  //   constraints: {
+  //     scale: function(scale) {
+  //       return /(celsius|fahrenheit|kelvin)/.test(scale);
+  //     }
+  //   }
+  // });
+});
+```
+
+### Combination of optional segments with constraints
+
+Both approaches can be combined. Imagine that in the weather forecast example, if the there is
+no temperature scale specified, the app will detect the best one based on the location of the
+device.
+
+On that situation, `/forecast` is a valid URL and has to point to the same route as the others with
+specific scale.
+
+```js
+const allowedScales = ['celsius', 'fahrenheit', 'kelvin'];
+
+function isAllowedScale(scaleName) {
+  return allowedScales.includes(scaleName);
+}
+
+Router.map(function() {
+  this.route('forecast', { path: 'forecast(/:scale)', constraints: { scale: isAllowedScale } });
+});
+
+
+router.serialize('/forecast/celsius'); // => routeName='forecast', params: { scale: 'celsius' }
+router.serialize('/forecast/newton'); // => not recognized route.
+router.serialize('/forecast'); // => routeName='forecast', params: { }
+```
+
+On this example `forecast` route is invoked even if the scale is ommited, but if it's not ommited,
+the constraint ensures that it's among a white-list of supported scales.
```

Comment:
Further, the `scale` function is invoked only if there is a segment following `/forecast` to identify it could possibly match the optional dynamic path segment.